File: App/ms-zonas-espacios/src/main/java/ec/edu/espe/zonas/cache/RedisCacheService.java
package ec.edu.espe.zonas.cache;

import tools.jackson.core.type.TypeReference;
import tools.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.stereotype.Service;

import java.time.Duration;
import java.util.function.Supplier;
import ec.edu.espe.zonas.tenant.TenantContext;

/**
 * Cache basada en Redis para las lecturas de espacios.
 * Registra en logs: "Cache HIT", "Cache MISS", "Cache SET", "Cache EVICT".
 * Si Redis no esta disponible, degrada de forma transparente (consulta directa a la BD).
 */
@Service
public class RedisCacheService {

    private static final Logger log = LoggerFactory.getLogger("RedisCache");

    private final StringRedisTemplate redis;
    private final ObjectMapper mapper;
    private final long defaultTtl;

    public RedisCacheService(StringRedisTemplate redis,
                             ObjectMapper mapper,
                             @Value("${redis.ttl:60}") long defaultTtl) {
        this.redis = redis;
        this.mapper = mapper;
        this.defaultTtl = defaultTtl;
    }

    /**
     * Devuelve el valor de cache si existe (HIT); si no, ejecuta {@code factory},
     * cachea el resultado (SET) y lo devuelve (MISS).
     */
    public <T> T getOrSet(String key, TypeReference<T> type, Supplier<T> factory) {
        return getOrSet(key, type, factory, defaultTtl);
    }

    public <T> T getOrSet(String key, TypeReference<T> type, Supplier<T> factory, long ttlSeconds) {
        key = TenantContext.get() + ":" + key;
        try {
            String cached = redis.opsForValue().get(key);
            if (cached != null) {
                log.info("Cache HIT  {}", key);
                return mapper.readValue(cached, type);
            }
            log.info("Cache MISS {}", key);
        } catch (Exception e) {
            // Redis caido o error de parseo: seguimos sin cache
            log.warn("Redis no disponible para GET {} ({})", key, e.getMessage());
        }

        T value = factory.get();

        if (value != null) {
            try {
                redis.opsForValue().set(key, mapper.writeValueAsString(value), Duration.ofSeconds(ttlSeconds));
                log.info("Cache SET  {} (ttl={}s)", key, ttlSeconds);
            } catch (Exception e) {
                // No se pudo cachear, devolvemos el valor igualmente
                log.warn("Redis no disponible para SET {} ({})", key, e.getMessage());
            }
        }

        return value;
    }

    /** Invalida una clave (se llama cuando cambia algun espacio). */
    public void evict(String key) {
        key = TenantContext.get() + ":" + key;
        try {
            redis.delete(key);
            log.info("Cache EVICT {}", key);
        } catch (Exception e) {
            log.warn("Redis no disponible para EVICT {} ({})", key, e.getMessage());
        }
    }
}

File: App/ms-tickets/src/tickets/common/redis-cache.service.ts
import { Injectable, Logger, OnModuleDestroy } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import Redis from 'ioredis';

/**
 * Cache basada en Redis para las validaciones de persona/vehiculo de ms-tickets.
 * Registra en logs: "Cache HIT", "Cache MISS", "Cache SET".
 * Si Redis no esta disponible, degrada de forma transparente (consulta directa).
 */
@Injectable()
export class RedisCacheService implements OnModuleDestroy {
  private readonly logger = new Logger('RedisCache');
  private readonly client: Redis;
  private readonly defaultTtl: number;
  private conectado = false;

  constructor(config: ConfigService) {
    const host = config.get<string>('REDIS_HOST') || 'localhost';
    const port = parseInt(config.get<string>('REDIS_PORT') || '6379', 10);
    this.defaultTtl = parseInt(config.get<string>('REDIS_TTL') || '60', 10);

    this.client = new Redis({
      host,
      port,
      // Fallar rapido si Redis no responde, en vez de encolar comandos
      enableOfflineQueue: false,
      maxRetriesPerRequest: 1,
      retryStrategy: (times) => Math.min(times * 300, 3000),
    });

    this.client.on('ready', () => {
      if (!this.conectado) this.logger.log(`Conectado a Redis en ${host}:${port}`);
      this.conectado = true;
    });
    this.client.on('error', (e) => {
      if (this.conectado) this.logger.warn(`Redis no disponible: ${e.message}`);
      this.conectado = false;
    });
  }

  /**
   * Devuelve el valor de cache si existe (HIT); si no, ejecuta `factory`,
   * cachea el resultado (SET) y lo devuelve (MISS).
   * Los errores de `factory` se propagan y NO se cachean.
   */
  async getOrSet<T>(tenantId: string, key: string, factory: () => Promise<T>, ttlSeconds?: number): Promise<T> {
    key = `${tenantId}:${key}`;
    const ttl = ttlSeconds ?? this.defaultTtl;

    try {
      const cached = await this.client.get(key);
      if (cached !== null) {
        this.logger.log(`Cache HIT  ${key}`);
        return JSON.parse(cached) as T;
      }
      this.logger.log(`Cache MISS ${key}`);
    } catch {
      // Redis caido: seguimos sin cache
    }

    const value = await factory();

    if (value !== null && value !== undefined) {
      try {
        await this.client.set(key, JSON.stringify(value), 'EX', ttl);
        this.logger.log(`Cache SET  ${key} (ttl=${ttl}s)`);
      } catch {
        // Redis caido: no se cachea, pero devolvemos el valor igualmente
      }
    }

    return value;
  }

  /** Invalida una clave (util cuando cambia el recurso). */
  async del(key: string): Promise<void> {
    try {
      await this.client.del(key);
    } catch {
      /* noop */
    }
  }

  onModuleDestroy(): void {
    this.client.quit().catch(() => this.client.disconnect());
  }
}
